image-registry/remove-registry-baremetal.md

```diff
@@ -0,0 +1,137 @@
+  - "@derekwaynecarr"
+creation-date: 2019-09-25
+last-updated: 2019-09-25
+status: provisional
+see-also: 
+replaces:
+superseded-by:
+---
+
+# Remove Registry on Baremetal
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift/docs]
+
+## Summary
+
+On `baremetal` platform clusters, the OpenShift image registry operator will bootstrap itself as `Removed`.
+This will allow `openshift-installer` to complete installations on this platform type.
+A separate process will be responsible for configuring storage and updating the image registry operator's state to `Managed`.
+
+## Motivation
+
+IPI installation options for Kubernetes-native infrastructure (formerly RHHI.next) and Red Hat Virtualization (RHV) have been declared a key initiative for OpenShift 4.3.
+These will configure themselves as the `baremetal` platform type.
+
+In 4.2 the registry bootstraps itself to use `EmptyDir` storage on `baremetal` to enable pre-GA developemnt on these platforms ([PR #332](https://github.com/openshift/cluster-image-registry-operator/pull/332)).
+This is not suitable for production as there is a serious risk of data loss if a permanent form of storage is not configured.
+
+### Goals
+
+- `openshift-installer` completes if the cloud provider is set to `baremetal`.
+- Image registry is able to bring itself up after the management state is switched from `Removed` to `Managed`.
+
+### Non-Goals
+
+- Provision storage for the registry for RHV and k8s-native infrastructre.
+- Provide 1st class support for additional storage types/providers for the registry (ex: MCG).
+- Harden OpenShift features (ex: Builds, ImageStreams) such that they error cleanly if the registry is not installed.
+
+## Proposal
+
+This is where we get down to the nitty gritty of what the proposal actually is.
+
+### User Story
+
+As an OpenShift cluster admin/installer
+I want the registry to be `Removed` on RHV and k8s-native infra installations
+So that I can provision other storage providers
+And later enable the image registry with the desired storage
+
+### Implementation Details/Notes/Constraints
+
+- When the image registry operator bootstraps itself, it will inspect the cluster infrastructure platform status.
+- If the platform status type is `baremetal`, the registry will mark itself `Removed`.
+- A separate installation process will be resposible for provisioning storage. For RHV and k8s-native infra, this is expected to be a PVC with the `RWX` access mode.
+- Telemeter should be aware if a cluster's registry has been removed:
+  - Reason codes in the ClusterOperator conditions (`Available` in particular).
+  - Prometheus alerts warning registry has been removed.
+
+### Risks and Mitigations
+
+#### Image registry is never installed
+
```

Comment:
An alternative that might be more useful is a mode where the registry only supports proxying and does not allow pushes (if storage is not configured).  That then preserves import and pull through and resiliency.

I expect in 4.4 or 4.5 we will start looking at always having the registry proxy content in order to survive the failure of an external registry - it may be that configuring the registry in proxy mode is desirable even *without* mirroring in order to preserve that.